File: fixtures/mllike/file2.ml
let validate_input value =
  if String.length value = 0 then
    Error "Input cannot be empty"
  else if String.length value > 255 then
    Error "Input exceeds maximum length"
  else
    Ok value

let parse_int_safe s =
  try Some (int_of_string s)
  with Failure _ -> None

let group_by key_fn items =
  List.fold_left (fun acc item ->
    let key = key_fn item in
    let existing = try List.assoc key acc with Not_found -> [] in
    (key, item :: existing) :: List.remove_assoc key acc
  ) [] items

let retry max_attempts action =
  let rec loop attempt =
    if attempt >= max_attempts then Error "Max retries exceeded"
    else
      match action () with
      | Ok v    -> Ok v
      | Error _ -> loop (attempt + 1)
  in
  loop 0

(* File-specific: product type *)
type product = { id: int; name: string; price: float }

let make_product id name price = { id; name; price }

File: fixtures/mllike/file2.fs
module DataUtils

open System
open System.Collections.Generic

let validateInput (value: string) =
    if String.IsNullOrWhiteSpace(value) then
        Error "Input cannot be empty"
    elif value.Length > 255 then
        Error "Input exceeds maximum length"
    else
        Ok value

let parseIntSafe (s: string) =
    match Int32.TryParse(s) with
    | true, v  -> Some v
    | false, _ -> None

let groupBy (keyFn: 'a -> 'b) (items: 'a list) : Map<'b, 'a list> =
    List.fold (fun acc item ->
        let key = keyFn item
        let existing = Map.tryFind key acc |> Option.defaultValue []
        Map.add key (item :: existing) acc
    ) Map.empty items

let retryAsync (maxAttempts: int) (action: unit -> Async<Result<'a, string>>) =
    async {
        let mutable attempt = 0
        let mutable result = Error "Not started"
        while attempt < maxAttempts && result = Error "Not started" do
            let! r = action()
            result <- r
            attempt <- attempt + 1
        return result
    }

// File-specific: product record type
type Product = { Id: int; Name: string; Price: float }

let createProduct id name price : Product =
    { Id = id; Name = name; Price = price }
